classes/class-truelayer-config-editor.php

```diff
@@ -0,0 +1,163 @@
+	 * @param string $key_value The value of the key to add.
+	 *
+	 * @return bool True if the key was added, false if it was not.
+	 */
+	public function add_key( $key_name, $key_value ) {
+		$key_name  = $this->sanitize_key_name( $key_name );
+		$key_value = $this->sanitize_key_value( $key_value );
+
+		// Create a backup for the config file first.
+		$backup_result = $this->create_backup();
+
+		// If the backup failed, then we cannot continue, since it might mean the write protection of the folder is on.
+		if ( ! $backup_result ) {
+			return false; // Backup failed
+		}
+
+		$config_data = file( $this->config_path, FILE_IGNORE_NEW_LINES );
+
+		// Check if key already exists
+		foreach ( $config_data as $line ) {
+			if ( strpos( $line, $key_name ) !== false ) {
+				return false; // Key already exists
+			}
+		}
+
+		// Find the place to add the new key
+		$insert_at = 0;
+		foreach ( $config_data as $index => $line ) {
+
+			// Try to insert the new key after the nonce salt definition
+			if ( strpos( $line, 'NONCE_SALT' ) !== false ) {
+				$insert_at = $index + 1;
+				break;
+			}
+
+			// If we did not find the nonce salt definition, then test against the "stop editing" comment
+			if ( strpos( $line, '/* That\'s all, stop editing! Happy publishing. */' ) !== false ) {
+				$insert_at = $index;
+				break;
+			}
+		}
+
+		// Insert the new key
+		array_splice( $config_data, $insert_at, 0, "define( '$key_name', '$key_value' );" );
+
+		return $this->save_config( $config_data );
+	}
+
+	/**
+	 * Remove a specific key from the wp-config.php file.
+	 *
+	 * @param string $key_name The name of the key to remove.
+	 *
+	 * @return bool True if the key was removed, false if it was not.
+	 */
+	public function remove_key( $key_name ) {
+		$config_data = file( $this->config_path, FILE_IGNORE_NEW_LINES );
+
+		// Find the key and remove it
+		foreach ( $config_data as $index => $line ) {
+			if ( strpos( $line, $key_name ) !== false ) {
+				unset( $config_data[ $index ] );
+				return $this->save_config( $config_data );
+			}
+		}
+
+		return false; // Key not found
+	}
+
+	/**
+	 * Saves the config data to the wp-config.php file.
+	 *
+	 * @param array $config_data The config data to save.
+	 *
+	 * @return bool True if the config was saved, false if it was not.
+	 */
+	protected function save_config( $config_data ) {
+		$temp_file = $this->config_path . '.tmp';
+
+		// Write to a temporary file first
+		$write_success = file_put_contents( $temp_file, implode( PHP_EOL, $config_data ) );
+
+		if ( $write_success === false ) {
+			return false; // Writing to temp file failed
+		}
+
+		// Now rename the temporary file to replace the original config
+		return rename( $temp_file, $this->config_path );
+	}
+
+	/**
+	 * Create a backup of the wp-config.php file.
+	 *
+	 * @return bool True if the backup was created, false if it was not.
+	 */
+	protected function create_backup() {
+		return copy( $this->config_path, $this->backup_path );
```

Comment:
Slightly worried that if somnething goes wrong with writing to the config file and they retry, the backup file will be written to multiple times, essentially overwriting the original version with a potentially broken one.

Reply:
That is true, however this will only run once on either the initial installation or after you update to this version of the plugin.

Before we ran the key generation on the `init` hook in WordPress, which technically is triggered all the time. If we had that logic this would have been a bigger potential issue.

Right now this method is triggered in the migration function, which should only run once since we both store it in the database that the version has been migrated to, as well as the specific function has already been processed. So for this method to run twice we would need two database options to be removed, as well as the key to not be defined.

With that said, we could timestamp the name or add a random string to it if you want to add extra security. Do you want me to do that @lighe?

Reply:
Ah ok, I agree the risk is much lower then. If it's not going to add a lot of delays I think timestamping is a good idea. But also happy to do that in a separate release.